[SPARK-27610][YARN] Shade netty native libraries

amuraru opened the pull request:
## What changes were proposed in this pull request?

Fixed the `spark-<version>-yarn-shuffle.jar` artifact packaging to shade the native netty libraries:
- shade the `META-INF/native/libnetty_*` native libraries when packagin
the yarn shuffle service jar. This is required as netty library loader
derives that based on shaded package name.
- updated the `org/spark_project` shade package prefix to `org/sparkproject`
(i.e. removed underscore) as the former breaks the netty native lib loading.

This was causing the yarn external shuffle service to fail 
when spark.shuffle.io.mode=EPOLL

## How was this patch tested?
Manual tests

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

vanzin commented:
How does this work? Won't the symbols in the native libraries still match the original, non-relocated names?

e.g. in the source for the native lib I see this:

```
static jint netty_epoll_native_JNI_OnLoad(JNIEnv* env, const char* packagePrefix) {
    // We must register the statically referenced methods first!
    if (netty_unix_util_register_natives(env,
            packagePrefix,
            "io/netty/channel/epoll/NativeStaticallyReferencedJniMethods",
            statically_referenced_fixed_method_table,
            statically_referenced_fixed_method_table_size) != 0) {
        return JNI_ERR;
    }
    // Register the methods which are not referenced by static member variables
    JNINativeMethod* dynamicMethods = createDynamicMethodsTable(packagePrefix);
    if (netty_unix_util_register_natives(env,
            packagePrefix,
            "io/netty/channel/epoll/Native",
            dynamicMethods,
            dynamicMethodsTableSize()) != 0) {
        freeDynamicMethodsTable(dynamicMethods);
        return JNI_ERR;
    }
```

The shade plugin doesn't change those strings to be in the new relocated package. Never mind how the JVM is even able to match the relocate Java names to the symbols in the native library in the first place.

vanzin commented:
Hmmm... the netty native code seems to map the native symbols to java names based on the file name of the native library (see `netty_unix_util_parse_package_prefix`). Clever.

Usual requests:
- your PR title doesn't follow the convention (missing the [module] part)
- describe the fix, not the problem, in the PR title.

Also since it's a maven change you need to add `[test-maven]` to the PR title to run the build through maven.

vanzin commented:
I'm still a little at a loss about how this works. Following the netty code, given the file name of the library, the translated name for e.g. `io/netty/channel/epoll/Native` will be `org/spark/project/io/netty/channel/epoll/Native`, because of this code:

```
    // Package names must be sanitized, in JNI packages names are separated by '/' characters.
    for (; temp != packageNameEnd; ++temp) {
        if (*temp == '_') {
            *temp = '/';
        }
    }
```

So the native library should fail to register because that class does not exist (see `netty_unix_util_register_natives`).

That shouldn't be hard to fix (just change the relocated name to not have an underscore), but it makes me wonder how you're verifying that this fix actually works.

amuraru commented:
> That shouldn't be hard to fix (just change the relocated name to not have an underscore)

you're right, thanks for spotting that @vanzin  - I was testing with a different `spark.shade.packageName` value override in my environment

amuraru commented:
> * describe the fix, not the problem, in the PR title.

I am not sure I get this, isn't the commit/PR title supposed to match JIRA title?
And JIRA title is filed as a bug, so problem is described in the title.
I find that more useful than say `Shade netty native libraries in yarn shuffle service uber jar`

srowen commented on common/network-yarn/pom.xml:
```diff
@@ -35,7 +35,7 @@
     <hadoop.deps.scope>provided</hadoop.deps.scope>
     <shuffle.jar>${project.build.directory}/scala-${scala.binary.version}/spark-${project.version}-yarn-shuffle.jar</shuffle.jar>
-    <shade>org/spark_project/</shade>
+    <shade>org/spark/project/</shade>
```
I think this was on purpose, to avoid packaging into the org.spark namespace

vanzin replied:
Probably doesn't matter much anyway, but to be more consistent with the original intent, `org/sparkproject` would be better.

srowen replied:
I recall we tried `spark-project` at some point and it was valid in bytecode but not in Java. `spark_project` ought to be a valid part of a package. Weak. OK I see the argument for `org.sparkproject`

vanzin replied:
`spark_project` is valid, but you need to go and read the discussion to know why the change is needed. (Or the updated PR description which also mentions it.)

amuraru replied:
Makes sense @srowen @vanzin - updated the patch to shade in `org.sparkproject`, ie no underscore

vanzin commented:
> I am not sure I get this, isn't the commit/PR title supposed to match JIRA title?

No. Bugs explain problems. PRs explain fixes. If you want to know more about the problem, look at the bug.

vanzin commented on conf/log4j.properties.template:
```diff
@@ -28,8 +28,8 @@ log4j.appender.console.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss} %p %c{1}:
 # Settings to quiet third party logs that are too verbose
-log4j.logger.org.spark_project.jetty=WARN
-log4j.logger.org.spark_project.jetty.util.component.AbstractLifeCycle=ERROR
+log4j.logger.org.spark.project.jetty=WARN
```
None of these other files should be changed.

amuraru replied:
Can you elaborate? `spark.shade.packageName` parent parameter is used across the entire project. 
Are you suggesting to have different shade prefix for `common/network-yarn` module only?

vanzin commented on pom.xml:
```diff
@@ -202,7 +202,7 @@
 
     <!-- Package to use when relocating shaded classes. -->
-    <spark.shade.packageName>org.spark_project</spark.shade.packageName>
+    <spark.shade.packageName>org.spark.project</spark.shade.packageName>
```
This one doesn't need to change.

amuraru replied:
can you elaborate @vanzin  - `common/network-yarn` package uses this property from parent pom to get the shade prefix

vanzin replied:
Hmm, I see what the other one is doing now. My bad.

amuraru committed:
[SPARK-27610][YARN] Shade netty native libraries


- shade the `META-INF/native/libnetty_*` native libraries when packagin
the yarn shuffle service jar. This is required as netty library loader
derives that based on shaded package name.
- updated the `org/spark_project` shade package prefix to `org/sparkproject`
(i.e. removed underscore) as the former breaks the netty native lib loading.

This was causing the yarn external shuffle service to fail 
when spark.shuffle.io.mode=EPOLL

amuraru commented:
> No. Bugs explain problems. PRs explain fixes. If you want to know more about the problem, look at the bug.

I see your point and the note in the Contributing Guide:
```The PR title should be of the form [SPARK-xxxx][COMPONENT] Title, where SPARK-xxxx is the relevant JIRA number, COMPONENT is one of the PR categories shown at spark-prs.appspot.com and Title may be the JIRA’s title or a more specific title describing the PR itself.```

I personally find it more consistent and easier to search (from a user perspective) the git history when the commit title describes the problem and not necessarily the fix, which can be described in the commit body.
Anyway, updated the commit and PR title per your suggestion.

vanzin commented:
> git history when the commit title describes the problem

At the risk of digressing too much, I find it extremely hard to understand a PR that explains what the problem is instead of explaining what it is doing. Many, many times the bug description is not a good way to understand the PR that fixes it.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10375/
Test PASSed.

SparkQA commented:
**[Test build #105084 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105084/testReport)** for PR 24502 at commit [`6886ae0`](https://github.com/apache/spark/commit/6886ae047359537986a48bd61fd4e3a1dfbc394e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10376/
Test PASSed.

SparkQA commented:
**[Test build #105084 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105084/testReport)** for PR 24502 at commit [`6886ae0`](https://github.com/apache/spark/commit/6886ae047359537986a48bd61fd4e3a1dfbc394e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105084/
Test PASSed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10382/
Test PASSed.

SparkQA commented:
**[Test build #105092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105092/testReport)** for PR 24502 at commit [`6886ae0`](https://github.com/apache/spark/commit/6886ae047359537986a48bd61fd4e3a1dfbc394e).

SparkQA commented:
**[Test build #105092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105092/testReport)** for PR 24502 at commit [`6886ae0`](https://github.com/apache/spark/commit/6886ae047359537986a48bd61fd4e3a1dfbc394e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105092/
Test FAILed.

amuraru committed:
Fixed MiMa tests

keep old spark_project in exclusion list for a while

vanzin commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10399/
Test PASSed.

SparkQA commented:
**[Test build #105110 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105110/testReport)** for PR 24502 at commit [`218a0e6`](https://github.com/apache/spark/commit/218a0e6d8183f6d0e5eb72061b97a5dd0239486a).

SparkQA commented:
**[Test build #105110 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105110/testReport)** for PR 24502 at commit [`218a0e6`](https://github.com/apache/spark/commit/218a0e6d8183f6d0e5eb72061b97a5dd0239486a).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105110/
Test FAILed.

vanzin commented:
It might be less work to override the shaded package name in the yarn shuffle service packaging...

amuraru committed:
Fixed MiMa tests take 2

amuraru commented:
one last try to fix Mima.
@vanzin if this won't fix the issue - I can try to constrain the fix in the yarn package only

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10419/
Test PASSed.

SparkQA commented:
**[Test build #105130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105130/testReport)** for PR 24502 at commit [`311d98f`](https://github.com/apache/spark/commit/311d98f22bd5ffa5e3961043d4848a6483869b61).

SparkQA commented:
**[Test build #105130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105130/testReport)** for PR 24502 at commit [`311d98f`](https://github.com/apache/spark/commit/311d98f22bd5ffa5e3961043d4848a6483869b61).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105130/
Test PASSed.

vanzin commented:
One reason why I have a preference for the yarn-packaging-only fix is that this is potentially messing people's logging configuration, in case they saw a need to customize Jetty logs. I don't think that's a huge deal though.

Anyway, will leave it here until tomorrow to see if anyone else has comments.

vanzin commented:
Merging to master.
